### PR TITLE
[CELEBORN-494][PERF] RssInputStream fetch side support blacklist to avoid client side timeout in same worker multiple times during fetch

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
####Test case
```
executor instance 20

SQL:
SELECT count(1) from (select /*+ REPARTITION(100) */ * from spark_auxiliary.t50g) tmp;


create connection timeout 10s

Fetch chunk timeout 30s
```
In the graph, the shuffle read time of `before` and `after` is always the same delay time.

##### Worker can't connect 
Before 
image

After 
image

##### OpenStream stuck
Before
image

After
image

##### Fetch chunk stuck
Before
image

After
image

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @waitinfuture @RexXiong @FMX

**Comment by waitinfuture** [comment]
Thanks for the PR @AngersZhuuuu ! Test cases seems very positive in worker-unavailable scenarios. I'm thinking whether we should add mechanism to remove workers from blacklist, since we can't guarantee the worker is broken or caused by fluctuation.

**Comment by AngersZhuuuu** [comment]
> Test cases seems very positive in worker-unavailable scenarios. I'm thinking whether we should add mechanism to remove workers from blacklist, since we can't guarantee the worker is broken or caused by fluctuation.

For the same PartitionLocation, if the peer has failed, how about allowing the blacklisted machine to retry?

**Comment by waitinfuture** [comment]
> 

Shall we add a timeout for blacklisted worker?

**Comment by waitinfuture** [comment]
> > Test cases seems very positive in worker-unavailable scenarios. I'm thinking whether we should add mechanism to remove workers from blacklist, since we can't guarantee the worker is broken or caused by fluctuation.
> 
> For the same PartitionLocation, if the peer has failed, how about allowing the blacklisted machine to retry?

In this PR, if both workers are blacklist, then the fetch will fast fail, maybe it's not what we want. We can add timeout for each blacklisted worker, and only consider a worker unavailable if at least two(configurable) exceptions happens.

**Comment by AngersZhuuuu** [comment]
@waitinfuture How about current?

**Comment by waitinfuture** [comment]
Hi @AngersZhuuuu , this PR seems very necessary, it'll be very nice if you make it done : )

**Comment by pan3793** [comment]
It seems like we are mixing up the terms "excludedWorker" and "blacklist". 

First of all, I don't object "blacklist", but we should unify them.

There are some options I think:

1. use "blacklist" anywhere
2. use "exclusion" anywhere, even it does not mention excluding what, but it should be clear we are excluding bad workers
3. use "excludedWorker" or "workerExclusion" based on the context.

File: client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java [hunk 1]
```diff
@@ -226,11 +237,56 @@ private void moveToNextReader() throws IOException {
...
+        return false;
+      } else {
+        return true;
+      }
+    }
+
+    private boolean criticalCause(Exception e) {
+      boolean isConnectTimeout =
+          e instanceof IOException
+              && e.getMessage().startsWith("Connecting to")
```
Comment (pan3793) [inline client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java:268]: please guard `e.getMessage() != null`

**Comment by waitinfuture** [comment]
> 

+1, shall we do code refactor before further related PRs?

File: client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java (hunk 1 above)
Comment (AngersZhuuuu) [reply client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java:269]: > please guard `e.getMessage() != null`

Done

**Comment by AngersZhuuuu** [comment]
After this pr I will make a new pr to unify the blacklist and execluded related configuration name.
Any more suggestion? @pan3793 @waitinfuture

File: client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java [hunk 2]
```diff
@@ -226,11 +237,57 @@ private void moveToNextReader() throws IOException {
...
+          e instanceof CelebornIOException
+              && e.getCause() != null
+              && e.getCause() instanceof IOException;
+      return isConnectTimeout || rpcTimeout || fetchChunkTimeout;
+    }
+
     private PartitionReader createReaderWithRetry(PartitionLocation location) throws IOException {
       while (fetchChunkRetryCnt < fetchChunkMaxRetry) {
         try {
+          if (isBlacklisted(location)) {
```
Comment (RexXiong) [inline client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java:285]: If the two replica already in blacklist(by others). There may be no partition location for this input stream can retry.

File: client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java [hunk 3]
```diff
@@ -399,6 +461,7 @@ public void close() {
         currentReader.close();
         currentReader = null;
       }
+      fetchBlacklist.clear();
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java:464]: > For fetchBlacklist shared by all RssInputStreams, why one single rssInputStream can clear this?

Only work for one RssInputStreamImpl

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 4]
```diff
@@ -2825,6 +2828,22 @@ object CelebornConf extends Logging {
...
+      .categories("client")
+      .doc("Whether to enable shuffle client-side fetch blacklist of workers.")
+      .version("0.3.0")
+      .booleanConf
+      .createWithDefault(false)
+
+  val CLIENT_FETCH_EXCLUDED_WORKER_EXPIRE_TIMEOUT: ConfigEntry[Long] =
+    buildConf("celeborn.client.fetch.exclude.expireTimeout")
+      .categories("client")
+      .doc("ShuffleClint is a static object, it will be used in the whole lifecycle of Executor," +
```
[the remark this answers is not in the document]
Comment (AngersZhuuuu) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:2842]: DOne

File: client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java (hunk 2 above)
Comment (AngersZhuuuu) [reply client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java:285]: Avoid this in `blacklistFailedLocation`

Comment (AngersZhuuuu) [reply client/src/main/java/org/apache/celeborn/client/read/RssInputStream.java:284]: > There may have three partitionLocation pair, such as A/B,B/C, A/C. If worker A And C exclude by first two, then if we try A/C, there is no partitionLocations left.

Change the behavior, how about current?

**Comment by AngersZhuuuu** [comment]
Any more suggestion? @pan3793 @RexXiong @waitinfuture

**Comment by waitinfuture** [comment]
LGTM, thanks! I'm thinking whether we should unify push blacklist and fetch blacklist. cc @AngersZhuuuu @pan3793 @RexXiong

**Pull request closed without merging**